Construct measurement timescales using control sys info

knelli2 opened the pull request:
## Proposed changes

Use control system info to construct the measurement timescales.

First commit factors out some of the work of constructing the initial expiration times to its own function. The measurement timescales will need the exact same code.
Second commit moves the `control_system::Tags::MeasurementTimescales` tag to its own file. (`ControlSystem/Tags.hpp` was a bit crowded and `control_system::Tags::FunctionsOfTimeInitialize` has its own file as well)

This also factors out a bit of work from #3612 to keep it as small as possible.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Add helper function for initial expiration times

wthrowe commented on src/ControlSystem/ConstructInitialExpirationTimes.hpp:
```diff
@@ -0,0 +1,43 @@
+#include "Utilities/TMPL.hpp"
+
+namespace control_system {
+struct ConstructInitialExpirationTimes {
```
Why is this a struct and not just a function?

knelli2 replied:
Brain fart.

nilsvu replied:
Looks like this got squashed into the wrong commit, please fix

nilsvu replied:
Also, note that we generally don't prepend verbs like "compute_" or "construct_" to function names when they can just be omitted, so I think `initial_expiration_times()` would be sufficient here.

wthrowe commented on src/ControlSystem/Tags/MeasurementTimescales.hpp:
```diff
@@ -0,0 +1,81 @@
+/// Each function of time associated with a control system has a corresponding
+/// set of timescales here, represented as `PiecewisePolynomial<0>` with the
+/// same components as the function itself.
+struct MeasurementTimescales : db::SimpleTag {
```
Move the test too.

knelli2 committed:
Move test for FunctionsOfTimeInitialize

knelli2 commented:
Added an extra commit that moves the test of FunctionsOfTimeInitialize to a new file to match the tree structure in the src directory.

wthrowe commented:
Squash.

knelli2 committed:
Move measurement timescales to their own file

No changes. Just moved the struct and test to a new file and replaced
includes where necessary

nilsvu commented on src/ControlSystem/ConstructInitialExpirationTimes.hpp:
```diff
@@ -0,0 +1,43 @@
+          const auto& tuner = option_holder.tuner;
+
+          const double update_fraction = controller.get_update_fraction();
+          const double curr_timescale = min(tuner.current_timescale());
```
#include DataVector

nilsvu commented on src/ControlSystem/ConstructInitialExpirationTimes.hpp:
```diff
@@ -0,0 +1,43 @@
+              update_fraction * curr_timescale;
+          initial_expiration_times[name] =
+              initial_time +
+              std::max(initial_time_step, initial_expiration_time);
```
#include `<algorithm>`

nilsvu commented on src/ControlSystem/Tags/FunctionsOfTimeInitialize.hpp:
```diff
@@ -50,24 +51,9 @@ struct FunctionsOfTimeInitialize : domain::Tags::FunctionsOfTime,
-              std::max(initial_time_step, initial_expiration_time);
-        };
-    EXPAND_PACK_LEFT_TO_RIGHT(gather_initial_expiration_times(option_holders));
+    std::unordered_map<std::string, double> initial_expiration_times =
```
const

nilsvu commented on src/ControlSystem/Tags.hpp:
```diff
@@ -3,27 +3,12 @@
 
 #pragma once
 
-#include <array>
```
Include `<cstddef>`, `<string>`, maybe more

nilsvu commented on src/ControlSystem/Averager.hpp:
```diff
@@ -116,6 +116,8 @@ class Averager {
   /// is `true`/`false`.
   bool using_average_0th_deriv_of_q() const { return average_0th_deriv_of_q_; };
+  /// Returns the averaging timescale fraction
+  double get_avg_timescale_frac() const { return avg_tscale_frac_; }
```
Also here you can omit the "get_".

nilsvu commented on src/ControlSystem/ConstructInitialExpirationTimes.hpp:
```diff
@@ -10,34 +10,29 @@
-  }
-};
+template <typename... OptionHolders>
+std::unordered_map<std::string, double> construct_initial_expiration_times(
```
Add a docstring please

nilsvu commented on src/ControlSystem/Tags/MeasurementTimescales.hpp:
```diff
@@ -35,46 +40,62 @@ struct MeasurementTimescales : db::SimpleTag {
+
+          const double update_fraction = controller.get_update_fraction();
+          const double averaging_fraction = averager.get_avg_timescale_frac();
+          const DataVector curr_timescale = tuner.current_timescale();
```
const DataVector&

nilsvu reviewed (changes requested):
squash directly

knelli2 committed:
Modify measurement timescale tag

Uses control system info to determine what to set the initial
measurement timescales to.

knelli2 commented:
Squashed in changes. Moved `construct_initial_expiration_times` to `initial_expiration_times` (along with file name as well). Also rebased on latest develop.
